### PR TITLE
feat: new rawBody type on SvelteKit req

### PR DESCRIPTION
**Pull request by jthegedus**
# Summary

Update to the latest SvelteKit request type.

Labelled as `feat` as this is breaking.

Fixes: #122 

# Details

The new type is `Uint8Array`, which we already get as a Buffer type (Node.js class which extends Uint8Array) from the `import('firebase-functions').https.Request` type, so this simplifies things a bit.

The assignment of `headers` is currently a type error:

```
Type 'IncomingHttpHeaders' is not assignable to type 'Headers'.
  Index signatures are incompatible.
    Type 'string | string[]' is not assignable to type 'string'.
      Type 'string[]' is not assignable to type 'string'.ts(2322)
internal.d.ts(18, 2): The expected type comes from property 'headers' which is declared here on type 'Incoming'
```

but looking at the `Headers` type in SvelteKit is seems this may not be an issue in future

```js
// kit/types/helper.d.ts

// TODO we want to differentiate between request headers, which
// always follow this type, and response headers, in which
// 'set-cookie' is a `string[]` (or at least `string | string[]`)
// but this can't happen until TypeScript 4.3
export type Headers = Record<string, string>;
```

so I think it is okay to leave the `IncomingHttpHeaders` type for now as it hasn't been causing issues for people so far.

Thoughts @benmccann?

# Tasks

- [x] use Uint8Array for rawBody
- [x] type render function
- [x] remove `string[]` from headers
- [x] unit test `toSvelteKit*` functions. Use requests generated from SvelteKit todos template?
  - [x] test `toSvelteKitRequest` with GET
  - [x] test `toSvelteKitRequest` with POST
- [x] fix upstream type error. `firebase-functions.https.request` type is wrong as `rawBody` is **optional** as shown in upstream Node.js Functions Framework https://github.com/GoogleCloudPlatform/functions-framework-nodejs/blob/e76b9f6c49bcf8c12d8206212cae84cfb931a077/src/invoker.ts#L44 - https://github.com/firebase/firebase-functions/pull/959
- [x] improve integration tests to build from SvelteKit Todo app
- [x] e2e test which makes a request to the Cloud Function & Hosting app using the Firebase Emulator in CI

### CODE REVIEW

File: src/files/handler.js [hunk 1]
```diff
@@ -7,40 +8,45 @@ init();
...
+	);
 
-	return response.writeHead(404).end();
-};
+	return {
+		// Incoming
+		method: request.method,
+		headers: request.headers,
+		rawBody: new Uint8Array(request.rawBody),
+		// Body: request.body, Why does SvelteKit not use body?
```
Comment (jthegedus) [inline src/files/handler.js:43]: SvelteKit types say there is also an optional `body:ParameterizedBody` field. Not sure what this is for or if we should populate it.

File: src/files/handler.js [hunk 2]
```diff
@@ -7,40 +8,45 @@ init();
...
-		rawBody: finalRawBody
-	});
-
-	if (rendered) {
-		const {status, headers, body} = rendered;
-		return response.writeHead(status, headers).end(body);
-	}
+/**
+ * @param {import('firebase-functions').https.Request} request
+ * @return {import('@sveltejs/kit/types/internal').Incoming}
```
Comment (jthegedus) [inline src/files/handler.js:29]: Would be good if the types for the `init` and `render` functions were exported properly as I think this is bad form

File: src/files/handler.js (hunk 1 above)
Comment (benmccann) [reply src/files/handler.js:43]: I would guess that SvelteKit populates it from `rawBody` and that the adapters don't need to though I haven't looked into it

File: src/files/handler.js (hunk 2 above)
Comment (benmccann) [reply src/files/handler.js:29]: Yeah, you shouldn't have to reference anything from `internal`. PR welcome if you'd like to take a stab at it. CC @ignatiusmb who's been fixing up the types. Thanks to both of you for all the help!

**Comment by benmccann** [comment]
Hmm. I'm not quite sure what that type TODO is about. It looks like Rich added it in https://github.com/sveltejs/kit/pull/445. We're on TypeScript 4.3 right now, so we should be able to fix it

**Comment by benmccann** [comment]
PR to fix that TODO here: https://github.com/sveltejs/kit/pull/2248

Not sure where the `IncomingHttpHeaders` type is coming from, but it sounds like its type definition might be a bit off. I'm guessing it doesn't really matter, but anything using it might be a little harder to use

File: src/files/handler.js (hunk 1 above)
Comment (jthegedus) [reply src/files/handler.js:43]: `params` will be removed, it was a holdover from when I was figuring out the available fields and types (Incoming & Location) obvs I later found the final type which omits params.

**Comment by jthegedus** [comment]
> PR to fix that TODO here: [sveltejs/kit#2248](https://github.com/sveltejs/kit/pull/2248)
> 
> I'm not sure why `IncomingHttpHeaders` accepts `string[]` values. Maybe it's only because they use [the same structure](https://nodejs.org/api/http.html#http_message_headers) for incoming and outgoing headers, but then I'm not sure why they have different types. I think it's only valid to have a `string`. If you have multiple values they're supposed to be comma-separated.

Right, so `Record<string, string>` is the final type I should be targetting. The TODO was **not** to expand the type to `Record<string, string|string[]>` :+1:

I am no expert of the Buffer/Uint8Array types, does `rawBody: new Uint8Array(request.rawBody)` seem okay to people. I couldn't find much on the web about creating a Uint8Array from a populated Buffer.

**Comment by jthegedus** [comment]
>I'm not sure why IncomingHttpHeaders accepts string[] values. Maybe it's only because they use the same structure for incoming and outgoing headers, but then I'm not sure why they have different types. I think it's only valid to have a string. If you have multiple values they're supposed to be comma-separated.

It seems the only `string[]` on `IncomingHttpHeaders` is `set-cookie`. Is this an expected field on a request?

File: src/files/handler.js (hunk 2 above)
Comment (jthegedus) [reply src/files/handler.js:29]: Started a discussion here - https://github.com/sveltejs/kit/issues/2249

**Comment by benmccann** [comment]
> The TODO was not to expand the type to Record<string, string|string[]>

Yeah. It was to change it to something like `Record<string, string> & { 'Set-Cookie'?: string | string[] };`. But that only sort of works sometimes even with TypeScript 4.3. I asked our liaison on the TypeScript team and he said it will probably never change

> I am no expert of the Buffer/Uint8Array types, does rawBody: new Uint8Array(request.rawBody) seem okay to people. I couldn't find much on the web about creating a Uint8Array from a populated Buffer.

I'm not an expert on this either. All I could do is check the other adapters and Google search it

> It seems the only string[] on IncomingHttpHeaders is set-cookie. Is this an expected field on a request?

No. I'm relatively sure you can only set that on responses

File: src/files/handler.js (hunk 2 above)
Comment (benmccann) [reply src/files/handler.js:29]: Cool. To answer @ignatiusmb's question, I haven't looked super closely, so am not 100% sure on this, but I feel like 90% sure that we just need to expose the types - though we should review and make sure it's all stuff we want to expose that would be my starting point

**Comment by jthegedus** [comment]
Great, thanks for the feedback and information @benmccann appreciate it.

**Comment by jthegedus** [comment]
Finally, the e2e test suite will:
- init a new SvelteKit todo template
- add the svelte-adapter-firebase & firebase config & functions/ dir with correct minimum config as in the docs
- hosts the built app on the Firebase Emulator in CI
- cURLs each of the pages `/`, `/about`, `/todos` and POSTs a new todo item to `/todos`

The below log shows the `/todos` SSR page rewriting to the Function, running its init code, hitting the Todo templates external API. It also shows the POST, listing it's requested resource as `todos.json` and the resultant payload showing successful creation of a new item.

```shell
[hosting] Rewriting /todos to http://localhost:5001/demo/us-central1/sveltekit for local Function sveltekit
i  functions: Beginning execution of "us-central1-sveltekit"
>  {"severity":"INFO","message":"Initialising SvelteKit SSR Handler"}
>  {"severity":"INFO","message":"SvelteKit SSR Handler initialised!"}
>  {"severity":"INFO","message":"Requested resource: /todos"}
⚠  External network resource requested!
   - URL: "api.svelte.dev/todos/bb7e4cad-bbd0-4326-ab27-f2e11796f478"
 - Be careful, this may be a production service.
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100  2991    0  2991    0     0   4762      0 --:--:-- --:--:-- --:--:--  4755
100  2991    0  2991    0     0   4762      0 --:--:-- --:--:-- --:--:--  4755
====> Test POST to '/todos' API
i  hosting: 127.0.0.1 - - [23/Aug/2021:12:07:28 +0000] "GET /todos HTTP/1.1" 200 - "-" "curl/7.68.0"
i  functions: Finished "us-central1-sveltekit" in ~1s
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

[hosting] Rewriting /todos.json to http://localhost:5001/demo/us-central1/sveltekit for local Function sveltekit
i  functions: Beginning execution of "us-central1-sveltekit"
>  {"severity":"INFO","message":"Requested resource: /todos.json"}
⚠  External network resource requested!
   - URL: "api.svelte.dev/todos/0a52e7d5-25d4-4b12-b307-38756d00bbcb"
 - Be careful, this may be a production service.
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
i  hosting: 127.0.0.1 - - [23/Aug/2021:12:07:29 +0000] "POST /todos.json HTTP/1.1" 201 - "http://localhost:8685/todos" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:91.0) Gecko/20100101 Firefox/91.0"
i  functions: Finished "us-central1-sveltekit" in ~1s
100   277    0   100  100   177    398    705 --:--:-- --:--:-- --:--:--  1103
{"uid":"f21fb8bfa277444190d6636331cd178c2929","created_at":1629720449095,"text":"asdf","done":false}
```

:tada:

If my new `rawBody` code is erroneous, at least we can update these tests to better capture why in future.

**Pull request merged**